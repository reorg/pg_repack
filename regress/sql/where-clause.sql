SET client_min_messages = warning;

-- Create a test table with some data
CREATE TABLE tbl_where_clause (
    id int PRIMARY KEY,
    value text,
    active boolean,
    deleted_at timestamp
);

-- Insert some test data
INSERT INTO tbl_where_clause VALUES (1, 'one', true, NULL);
INSERT INTO tbl_where_clause VALUES (2, 'two', true, NULL);
INSERT INTO tbl_where_clause VALUES (3, 'three', false, NULL);
INSERT INTO tbl_where_clause VALUES (4, 'four', false, NULL);
INSERT INTO tbl_where_clause VALUES (5, 'five', true, NULL);

-- Check initial data
SELECT * FROM tbl_where_clause ORDER BY id;

-- Run pg_repack with where clause to only repack active rows
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="active = true"

-- Verify data after repack
SELECT * FROM tbl_where_clause ORDER BY id;

-- Insert more data to verify the where clause worked correctly
INSERT INTO tbl_where_clause VALUES (6, 'six', true, NULL);
INSERT INTO tbl_where_clause VALUES (7, 'seven', false, NULL);

-- Run pg_repack with a different where clause
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="id > 3"

-- Verify data after second repack
SELECT * FROM tbl_where_clause ORDER BY id;

-- Test with where clause and order-by together
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="active = true" --order-by="value"

-- Verify data after repack with order-by
SELECT * FROM tbl_where_clause ORDER BY id;

INSERT INTO tbl_where_clause VALUES (8, 'eight', true, NULL);
INSERT INTO tbl_where_clause VALUES (9, 'nine', false, '2023-01-04 10:00:00');

-- Test with deleted_at is null where clause (keep only non-deleted rows)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="deleted_at IS NULL"

-- Verify data after repack with deleted_at IS NULL
SELECT * FROM tbl_where_clause ORDER BY id;

-- Test with non-existent column in where clause (should fail)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="non_existent_column = true"

-- Test with special character in column name without proper quoting (should fail)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="column with space = 'test'"

-- Test for repackaing whole database with a where clause (should fail)
\! pg_repack --dbname=contrib_regression --where-clause="id > 3"

-- Clean up
DROP TABLE tbl_where_clause; 