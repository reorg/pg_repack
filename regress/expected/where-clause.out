SET client_min_messages = warning;
-- Create a test table with some data
CREATE TABLE tbl_where_clause (
    id int PRIMARY KEY,
    value text,
    active boolean,
    deleted_at timestamp
);
-- Insert some test data
INSERT INTO tbl_where_clause VALUES (1, 'one', true, NULL);
INSERT INTO tbl_where_clause VALUES (2, 'two', true, NULL);
INSERT INTO tbl_where_clause VALUES (3, 'three', false, NULL);
INSERT INTO tbl_where_clause VALUES (4, 'four', false, NULL);
INSERT INTO tbl_where_clause VALUES (5, 'five', true, NULL);
-- Check initial data
SELECT * FROM tbl_where_clause ORDER BY id;
 id | value | active | deleted_at 
----+-------+--------+------------
  1 | one   | t      | 
  2 | two   | t      | 
  3 | three | f      | 
  4 | four  | f      | 
  5 | five  | t      | 
(5 rows)

-- Run pg_repack with where clause to only repack active rows
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="active = true"
INFO: repacking table "public.tbl_where_clause"
-- Verify data after repack
SELECT * FROM tbl_where_clause ORDER BY id;
 id | value | active | deleted_at 
----+-------+--------+------------
  1 | one   | t      | 
  2 | two   | t      | 
  5 | five  | t      | 
(3 rows)

-- Insert more data to verify the where clause worked correctly
INSERT INTO tbl_where_clause VALUES (6, 'six', true, NULL);
INSERT INTO tbl_where_clause VALUES (7, 'seven', false, NULL);
-- Run pg_repack with a different where clause
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="id > 3"
INFO: repacking table "public.tbl_where_clause"
-- Verify data after second repack
SELECT * FROM tbl_where_clause ORDER BY id;
 id | value | active | deleted_at 
----+-------+--------+------------
  5 | five  | t      | 
  6 | six   | t      | 
  7 | seven | f      | 
(3 rows)

-- Test with where clause and order-by together
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="active = true" --order-by="value"
INFO: repacking table "public.tbl_where_clause"
-- Verify data after repack with order-by
SELECT * FROM tbl_where_clause ORDER BY id;
 id | value | active | deleted_at 
----+-------+--------+------------
  5 | five  | t      | 
  6 | six   | t      | 
(2 rows)

INSERT INTO tbl_where_clause VALUES (8, 'eight', true, NULL);
INSERT INTO tbl_where_clause VALUES (9, 'nine', false, '2023-01-04 10:00:00');
-- Test with deleted_at is null where clause (keep only non-deleted rows)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="deleted_at IS NULL"
INFO: repacking table "public.tbl_where_clause"
-- Verify data after repack with deleted_at IS NULL
SELECT * FROM tbl_where_clause ORDER BY id;
 id | value | active | deleted_at 
----+-------+--------+------------
  5 | five  | t      | 
  6 | six   | t      | 
  8 | eight | t      | 
(3 rows)

-- Test with non-existent column in where clause (should fail)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="non_existent_column = true"
ERROR: invalid WHERE clause: column "non_existent_column" does not exist
-- Test with special character in column name without proper quoting (should fail)
\! pg_repack --dbname=contrib_regression --table=tbl_where_clause --where-clause="column with space = 'test'"
ERROR: invalid WHERE clause: syntax error at or near "column"
-- Test for repackaing whole database with a where clause (should fail)
\! pg_repack --dbname=contrib_regression --where-clause="id > 3"
ERROR: where-clause can only be used when repacking a single table.
-- Clean up
DROP TABLE tbl_where_clause; 
